### ISSUE TITLE
Using --filter="...[<sha>...<sha>]"` take local changes into account

### ISSUE THREAD

**Issue by EWhite613** [vercel/turbo#7818]
### Verify canary release

- [X] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

https://github.com/EWhite613/turbo-issue-cache/tree/test-change

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Which canary version will you have in your reproduction?

1.13.0-canary.1

### Describe the Bug

`turbo` takes into account local file changes when doing `turbo run test --filter="...[origin/main...HEAD]"`

### Expected Behavior

Expect local files and changes to not be taken into account. Should only be the commit shas directly referenced

### To Reproduce

```
git clone git@github.com:EWhite613/turbo-issue-cache.git
cd turbo-issue-cache
git checkout test-change
```

Run `turbo run test --filter="...[origin/main...HEAD]"` twice.  Notice first time it correctly only executes `api:test`. Second time it runs `api:test` and `@repo/logger:test` due to local files outputted (which are not in filter command)

### Additional context

Noticed this on both latest and canary.

Makes it difficult to only execute only a PRs changes (have to ensure I clean everything in between commands)

## Workaround
Specify `inputs`.

But I notice if I do this, I would have to do this for all my commands. Since say I have the following
```
{
  "$schema": "https://turbo.build/schema.json",
  "pipeline": {
    "test": {
      "dependsOn": ["^compile"],
      "outputs": ["artifacts/logs/test/xml/test-reports/xunit_results.xml"]
    },
    "lint": {
      "dependsOn": ["^compile", "^lint"]
    },
    "compile": {
      "dependsOn": ["^prepublishOnly"],
      "inputs": ["src/**"],
      "outputs": ["**/*.d.ts"]
    }
  }
}
```
I now need to let my test command be specific with it's inputs as well cause it will also see `compile`'s output as a change.

**Comment by EWhite613** [comment]
Note `turbo run test --filter="[origin/main...HEAD]"` seems to work as expected. So it's something with the including the dependents that changes this `...[<sha>...<sha>]`

**Comment by gsoltis** [comment]
This fix should be available starting at version `1.13.1-canary.3`
